### PR TITLE
Migrate configuration

### PR DESCRIPTION
**Pull request by shackra**
# Pull Request

To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [ ] Tests (if applicable)
* [ ] Documentation (if applicable)
* [ ] Changelog entry
* [ ] A full explanation here in the PR description of the work done

## PR Type
What kind of change does this PR introduce?

* [ ] Bugfix
* [ ] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the Connector in any way?

* [ ] Yes (backward compatible)
* [ ] No (breaking changes)

#### Useful links:

- [https://github.com/ConductorOne/baton-sdk/wiki/Coding-Guidelines](Baton SDK coding guidelines)
- [https://github.com/ConductorOne/baton/blob/main/CONTRIBUTING.md](New contributor guide)

## What's new?
-

### CODE REVIEW

File: cmd/baton-jenkins/config.go [hunk 1]
```diff
@@ -1,40 +1,20 @@
...
-type config struct {
-	cli.BaseConfig  `mapstructure:",squash"` // Puts the base config options in the same place as the connector options
-	JenkinsUsername string                   `mapstructure:"jenkins-username" description:"Username of administrator used to connect to the Jenkins API."`
-	JenkinsPassword string                   `mapstructure:"jenkins-password" description:"Application password used to connect to the Jenkins API."`
-	JenkinstBaseUrl string                   `mapstructure:"jenkins-baseurl" description:"Jenkins. example http://localhost:8080." defaultValue:"http://localhost:8080"`
-	JenkinsToken    string                   `mapstructure:"jenkins-token" description:"HTTP access tokens in Jenkins"`
+var relationships = []field.SchemaFieldRelationship{
+	field.FieldsMutuallyExclusive(token, username),
+	field.FieldsMutuallyExclusive(token, password),
+	field.FieldsRequiredTogether(username, password),
```
Comment (mgaeta) [inline cmd/baton-jenkins/config.go:17]: It looks like the only relationship is that `password` and `token` are mutually exclusive.

**Review by mgaeta — CHANGES_REQUESTED** [review]
https://www.jenkins.io/doc/book/system-administration/authenticating-scripted-clients/
I think the password and token are basically interchangeable 
```go
func WithSetBasicAuthHeader(username, password string) uhttp.RequestOption {
	return uhttp.WithHeader("Authorization", "Basic "+basicAuth(username, password))
}

func WithSetTokenAuthHeader(username, token string) uhttp.RequestOption {
	return uhttp.WithHeader("Authorization", "Basic "+basicAuth(username, token))
}
```
Also, make sure to update the help text in the README

File: cmd/baton-jenkins/main.go [hunk 2]
```diff
@@ -36,18 +36,18 @@ func main() {
 	}
 }
 
-func getConnector(ctx context.Context, cfg *config) (types.ConnectorServer, error) {
+func getConnector(ctx context.Context, v *viper.Viper) (types.ConnectorServer, error) {
 	l := ctxzap.Extract(ctx)
 	jenkinsClient := client.NewClient()
-	if cfg.JenkinsToken != "" {
-		jenkinsClient.WithUser(cfg.JenkinsUsername).WithBearerToken(cfg.JenkinsToken)
+	if v.GetString("token") != "" {
```
Comment (mgaeta) [inline cmd/baton-jenkins/main.go:42]:
```go
l := ctxzap.Extract(ctx)
err := field.Validate(Configuration, v)
if err != nil {
  return nil, err
}
jenkinsClient := client.NewClient().
  WithUser(v.GetString("username")).
  WithPassword(v.GetString("password")).
  WithToken(v.GetString("token"))
```

**Comment by shackra** [comment]
> https://www.jenkins.io/doc/book/system-administration/authenticating-scripted-clients/ I think the password and token are basically interchangeable
> 
> ```go
> func WithSetBasicAuthHeader(username, password string) uhttp.RequestOption {
> 	return uhttp.WithHeader("Authorization", "Basic "+basicAuth(username, password))
> }
> 
> func WithSetTokenAuthHeader(username, token string) uhttp.RequestOption {
> 	return uhttp.WithHeader("Authorization", "Basic "+basicAuth(username, token))
> }
> ```
> 
> Also, make sure to update the help text in the README

the client API for Jenkins is a bit different so we cannot re-use variables to set that piece of auth information:

```go
jenkinsClient.WithUser("random-username").WithBearerToken("magic-token")

jenkinsClient.WithUser("random-username").WithPassword("secretpassword")
```

File: cmd/baton-jenkins/config.go [hunk 3]
```diff
@@ -1,40 +1,18 @@
...
 
-// config defines the external configuration required for the connector to run.
-type config struct {
-	cli.BaseConfig  `mapstructure:",squash"` // Puts the base config options in the same place as the connector options
-	JenkinsUsername string                   `mapstructure:"jenkins-username" description:"Username of administrator used to connect to the Jenkins API."`
-	JenkinsPassword string                   `mapstructure:"jenkins-password" description:"Application password used to connect to the Jenkins API."`
-	JenkinstBaseUrl string                   `mapstructure:"jenkins-baseurl" description:"Jenkins. example http://localhost:8080." defaultValue:"http://localhost:8080"`
-	JenkinsToken    string                   `mapstructure:"jenkins-token" description:"HTTP access tokens in Jenkins"`
+var relationships = []field.SchemaFieldRelationship{
+	field.FieldsMutuallyExclusive(token, password),
```
Comment (mgaeta) [inline cmd/baton-jenkins/config.go:15]: I think you should also add a `FieldsAtLeastOneUsed` relationship for `token` and `password`

**Review by mgaeta — APPROVED** [review]

File: cmd/baton-jenkins/main.go (hunk 2 above)
Comment (shackra) [reply cmd/baton-jenkins/main.go:42]: I mean, is that even possible without causing API errors? 😆

**Pull request merged**